Add color palette

sophiiae opened the pull request.

natashadecoste commented on packages/core/src/styles/colors.scss:
```diff
@@ -0,0 +1,59 @@
+}
+
+$name: 'color-fill';
+$tokens: (
```
are you going to map tokens to this variable based on the number of datagroups?

natashadecoste commented on packages/core/src/styles/colors.scss:
```diff
@@ -0,0 +1,59 @@
+	$current-theme: "light";
+}
+
+$name: 'color-fill';
```
do you need name though? i dont think this part would change based on the no. of datasets, the palette chosen or the theme 🤔

theiliad commented on packages/core/src/components/graphs/area-stacked.ts:
```diff
@@ -67,11 +68,20 @@ export class StackedArea extends Component {
 
 		const enteringAreas = areas.enter().append("path").attr("opacity", 0);
+		const userProvidedScale = Tools.getProperty(options, "color", "scale");
+		const noProvidedColorScale = userProvidedScale === null || Object.keys(userProvidedScale).length === 0;
 
 		enteringAreas
 			.merge(areas)
 			.data(stackedData, (d) => d[0][groupMapsTo])
-			.attr("fill", (d) => self.model.getFillColor(d[0][groupMapsTo]))
+			.attr("class", "area")
+			.attr(
+				"class",
+				(d) => noProvidedColorScale
+				? `area ${this.model.getColorClass(d[0][groupMapsTo])}`
+				: "area"
+			)
+			.attr("fill", (d) => noProvidedColorScale ? null : self.model.getFillColor(d[0][groupMapsTo]))
```
Only `model` should be looking at `Tools.getProperty(options, "color", "scale")`

theiliad commented on packages/core/src/components/graphs/bubble.ts:
```diff
@@ -66,17 +69,27 @@ export class Bubble extends Scatter {
+				"class",
+				(d) => noProvidedColorScale
+					? `dot
+					${this.model.getStrokeColorClass(d[groupMapsTo])} 
+					${this.model.getColorClass(d[groupMapsTo])}
```
Can this be a `this.model.getColorClassNames(["fill", "stroke"], d[groupMapsTo])`?

theiliad commented on packages/core/src/components/graphs/gauge.ts:
```diff
@@ -121,14 +121,21 @@ export class Gauge extends Component {
 			.enter()
 			.append("path")
-			.attr("class", "arc-foreground")
+			.attr(
+				"class",
+				(d) => noProvidedColorScale
+					? `arc-foreground ${self.model.getColorClass(d[groupMapsTo])}`
+					: "arc-foreground"
+			)
+			.attr("fill", (d) => noProvidedColorScale ? null : self.model.getFillColor(d[groupMapsTo]))
```
we need to find a solution to this, it's not ideal to have every graph check for the custom color scale

theiliad commented on packages/core/src/components/graphs/radar.ts:
```diff
@@ -823,7 +834,7 @@ export class Radar extends Component {
 							label: datum[groupMapsTo],
 							value: datum[valueMapsTo],
-							color: self.model.getStrokeColor(datum[groupMapsTo])
+							class: self.model.getTooltipColorClass(datum[groupMapsTo])
```
is the tooltip color class name any different than the fill color class name?

theiliad commented on packages/core/src/configuration.ts:
```diff
@@ -154,6 +154,9 @@ const chart: BaseChartOptions = {
 	},
 	color: {
 		scale: null,
+		presetPalette: {
```
is this a pairing? then why not just call it `pairing`?

theiliad commented on packages/core/src/index.ts:
```diff
@@ -4,9 +4,4 @@ export * from "./charts";
-const defaultColors = DEFAULT;
-
-export { interfaces, configurations, defaultColors, colorPalettes };
+export { interfaces, configurations };
```
👍

theiliad commented on packages/core/src/interfaces/charts.ts:
```diff
@@ -104,6 +104,16 @@ export interface BaseChartOptions {
+		 * use a carbon dataviz preset color palette
+		 * put the index (selection of which variant)
+		 */
+		presetPalette?: {
```
```suggestion
		pairing?: number;
```

theiliad commented on packages/core/src/model.ts:
```diff
@@ -326,7 +329,8 @@ export class ChartModel {
 
-		this.setColorScale();
+		this.setUserProvidedColorScale();
+		this.setColorClasses();
```
We should call them all `classNames` instead of `classes`, much more readable

theiliad commented on packages/core/src/model.ts:
```diff
@@ -578,47 +595,97 @@ export class ChartModel {
+			"presetPalette",
+			"index"
+		);
+		const paletteAvailableChoices = {
```
```suggestion
		const availableColorPairings = {
```
?

theiliad commented on packages/core/src/model.ts:
```diff
@@ -578,47 +595,97 @@ export class ChartModel {
+			"14-color": 1
+		};
+		const numberOfDataGroups =
+			this.allDataGroups.length > 5 ? 14 : this.allDataGroups.length;
```
why would we use a fake `numberOfDataGroups` value if more than 5? do we need a comment here to describe the decision?

theiliad commented on packages/core/src/model.ts:
```diff
@@ -578,47 +595,97 @@ export class ChartModel {
+					index + 1
+				}`
+		);
+
```
is there a difference between `tooltip`, `fill` & `stroke` classnames?

jeanservaas commented:
@theiliad this is cool.

It would be really awesome to have controls to change the color pairings @mjabbink. Right now you can only change it through code and we'll have to plan to add a code tab to the data viz Palette page to provide guidance on how to switch pairings. 

Also @mjabbink we might want to make sure that the pairings we have listed 1st (i.e. the preset palette 1) are the colors we want to have as the defaults. The one and two colors sets are fine. But the 3 and especially 4 color sets are not the most attractive. The four-color one is defaulting to the pairing that has the most red in it right now. @theiliad this is my fault because this is how I organized them. 


### three color pairings

**current:**

<img width="331" alt="image" src="https://user-images.githubusercontent.com/43144260/94942698-a556aa00-04a4-11eb-99ee-d94556cf8ef7.png">

**better default:**

<img width="330" alt="image" src="https://user-images.githubusercontent.com/43144260/94942746-b43d5c80-04a4-11eb-9098-3ebb3cf58b43.png">



### four color pairings

**current:**
<img width="351" alt="image" src="https://user-images.githubusercontent.com/43144260/94942808-c8815980-04a4-11eb-9916-00a95acab82d.png">


**better default:**

<img width="320" alt="image" src="https://user-images.githubusercontent.com/43144260/94942585-835d2780-04a4-11eb-8d93-0fc46453f510.png">



I noticed something on bubble charts in Safari, when I brought my cursor over the live chart area the 30% opacity within the bubbles disappeared and did not return unless I refreshed

<img width="1010" alt="image" src="https://user-images.githubusercontent.com/43144260/94939898-c74e2d80-04a0-11eb-9d52-5c274cfa6e38.png">

The same is happening with scatter:

<img width="1003" alt="image" src="https://user-images.githubusercontent.com/43144260/94940200-31ff6900-04a1-11eb-8ed1-6794f3d13e6a.png">
Also on these thresholds... how did we come up with these colors? Are they just meant to be completely configurable? I also noticed the dots above the threshold have a much thicker stroke and are slightly larger, is that intentional?

<img width="995" alt="image" src="https://user-images.githubusercontent.com/43144260/94940059-fbc1e980-04a0-11eb-935c-1bf07296115f.png">


**Completely unrelated to this PR:**

Why are the x-axis labels at a 45 degree angle here?

<img width="1002" alt="image" src="https://user-images.githubusercontent.com/43144260/94940324-59563600-04a1-11eb-8058-5d0316e767d5.png">

mjabbink commented:
I agree the current option 1 in 4-color group should not be the default and in fact become option 3.

sophiiae commented:
@jeanservaas Thank you for pointing out the scatter and bubble chart issue. I have fixed this and updated the default colors as you mentioned. The thresholds dash line color is completely configurable. Those two dots stand out with given thresholds, I didn't make any change for the threshold component in this PR. I believe it's intentional. If the x-axis labels are time series, the labels are rotated with 45 degrees.

natashadecoste commented on packages/core/src/components/essentials/legend.ts:
```diff
@@ -64,6 +64,13 @@ export class Legend extends Component {
+		const paletteIndex = Tools.getProperty(
+			options,
+			"color",
+			"paring",
```
```suggestion
			"pairing",
```

not sure im fully for the naming "pairing" but either way this is the right spelling

natashadecoste commented on packages/core/demo/data/bar.ts:
```diff
@@ -90,6 +90,11 @@ export const simpleBarOptions = {
 		}
+	},
+	color: {
+		paring: {
```
```suggestion
		pairing: {
```

natashadecoste commented on packages/core/src/styles/colors.scss:
```diff
@@ -0,0 +1,313 @@
+			"12": map-get(map-get($ibm-color-map, "cyan"), 90),
+			"13": #8a3800,
+			//orange 70, update when available in carbon color
+				"14": map-get(map-get($ibm-color-map, "purple"), 50)
```
```suggestion
			"14": map-get(map-get($ibm-color-map, "purple"), 50)
```

natashadecoste commented on packages/core/src/styles/colors.scss:
```diff
@@ -0,0 +1,313 @@
+			"9": map-get(map-get($ibm-color-map, "magenta"), 50),
+			"10": #b28600,
+			//yellow 50, update when available in carbon color
+				"11": map-get(map-get($ibm-color-map, "teal"), 50),
```
```suggestion
			"11": map-get(map-get($ibm-color-map, "teal"), 50),
```

natashadecoste commented on packages/core/src/styles/colors.scss:
```diff
@@ -0,0 +1,313 @@
+			"9": map-get(map-get($ibm-color-map, "magenta"), 60),
+			"10": #d2a106,
+			//yellow 40, update when available in carbon color
+				"11": map-get(map-get($ibm-color-map, "teal"), 40),
```
```suggestion
			"11": map-get(map-get($ibm-color-map, "teal"), 40),
```

natashadecoste commented on packages/core/src/styles/colors.scss:
```diff
@@ -0,0 +1,313 @@
+			"12": map-get(map-get($ibm-color-map, "cyan"), 20),
+			"13": #ba4e00,
+			//orange 60, update when available in carbon color
+				"14": map-get(map-get($ibm-color-map, "purple"), 30)
```
```suggestion
			"14": map-get(map-get($ibm-color-map, "purple"), 30)
```

natashadecoste commented on packages/core/src/configuration.ts:
```diff
@@ -154,6 +154,9 @@ const chart: BaseChartOptions = {
 	},
 	color: {
 		scale: null,
+		paring: {
```
```suggestion
		pairing: {
```

natashadecoste commented on packages/core/src/interfaces/charts.ts:
```diff
@@ -104,6 +104,16 @@ export interface BaseChartOptions {
+		 * use a carbon dataviz preset color palette
+		 * put the index (selection of which variant)
+		 */
+		paring?: {
```
```suggestion
		pairing?: {
```

natashadecoste commented on packages/core/src/model.ts:
```diff
@@ -454,6 +464,35 @@ export class ChartModel {
+			return originalClassName;
+		}
+
+		const colorParingTag = this.colorClassNames(group);
```
```suggestion
		const colorPairingTag = this.colorClassNames(group);
```

natashadecoste commented on packages/core/src/model.ts:
```diff
@@ -454,6 +464,35 @@ export class ChartModel {
+			(type) =>
+				(className = originalClassName
+					? `${className} ${type}-${colorParingTag}`
+					: `${type}-${colorParingTag}`)
```
```suggestion
					: `${type}-${colorPairingTag}`)
```

natashadecoste commented on packages/core/src/model.ts:
```diff
@@ -584,47 +623,71 @@ export class ChartModel {
+		let paringIndex = Tools.getProperty(
+			this.getOptions(),
+			"color",
+			"paring",
```
```suggestion
			"pairing",
```

natashadecoste commented on packages/core/src/model.ts:
```diff
@@ -584,47 +623,71 @@ export class ChartModel {
+	 * Color palette
+	 */
+	protected setColorClasses() {
+		let paringIndex = Tools.getProperty(
```
```suggestion
		let pairingIndex = Tools.getProperty(
```

natashadecoste reviewed (changes requested):
it doesn't seems that you can choose a bigger palette than the number of datagroups.
What if i have 2 expected datagroups (ex days of the weekend) but I know that i might sometimes have 3 (a long weekend) and i'd like the colors to stay consistent in the chart/dashboard. 

could it make sense to allow someone to pick a 3 color palette for 2 datagroups in this case? Otherwise it seems the colors are going to jump around since there isnt a 2 color palette that extends to being a 3 color

natashadecoste commented on packages/core/src/interfaces/charts.ts:
```diff
@@ -104,6 +104,20 @@ export interface BaseChartOptions {
+			/**
+			 * the number of data groups/categories, only if differ from given data
+			 */
+			numberOfGroups?: number;
```
I'd rather this was called colorGroups, colorVariants or something that makes it more intuitive that its the number of **colors** rather than groups that you are overriding in the chart. Since i think you could just **choose** to use a bigger colorset for visual parity with the rest of the dashboard/page,  it makes this config seem a bit misleading to me.

sophiiae replied:
I know it's not a good name for this config, I'd like to see if anyone comes up with a better idea for this. I thought about `colorGroups`, it seems more like the input is an array of string, not an integer. It might be a bit confused about the type.

natashadecoste replied:
variants? numberOfVariants? 🤔

sophiiae replied:
I'll take `numberOfVariants`.

jeanservaas reviewed (approved):
I'm gonna approve this now because I really want to get it in, and then we can change the order of the color groupings, which will also need to be updated on the site and in the master charts file — into another PR.

natashadecoste commented on packages/core/src/interfaces/charts.ts:
```diff
@@ -104,6 +104,20 @@ export interface BaseChartOptions {
+		 */
+		pairing?: {
+			/**
+			 * the number of data groups/categories, only if differ from given data
```
```suggestion
			 * the number of color variants in the palette (defaults to using the number of data groups in the given data)
```

natashadecoste reviewed (changes requested):
What do we do if they are using custom colors? 
If i add custom color scale (on a line chart demo), it only colors those datasets -> it doesnt affect their **legend items**

![image](https://user-images.githubusercontent.com/14351335/96767261-4a7ce800-13aa-11eb-8976-c436c197c543.png)

And then there's the case of if they put less custom colors than there is datagroups - do we default them to carbon palette? or use a carbon palette for the colors that are remaining (not defined)? 

it turns out like this (below)
![image](https://user-images.githubusercontent.com/14351335/96767500-9e87cc80-13aa-11eb-8479-ce268f5ce12f.png)

same legend problem but also the chart **looks correct** but is mapping 2 datasets to the 2 colors when the chart needs 4 color variants. I argue this is worse than a chart warning or erroring out because its just looking really false

natashadecoste commented on packages/core/src/components/graphs/area.ts:
```diff
@@ -76,7 +76,17 @@ export class Area extends Component {
 
-		if (isGradientAllowed) {
+		// The fill value of area has been overwritten, get color value from stroke color class instead
+		const areaElement = document.getElementsByClassName(
+			this.model.getColorClassName(["stroke"], groupedData[0].name)
+		)[0];
```
why not use a d3 selection??

theiliad replied:
yes please update this

theiliad commented on packages/core/src/components/essentials/tooltip.ts:
```diff
@@ -174,15 +174,18 @@ export class Tooltip extends Component {
 		} else {
 			const items = this.getItems(e);
 			const formattedItems = this.formatItems(items);
+			const useColor = this.model.colorIsProvided(); 
```
we need a more readable name here

theiliad commented on packages/core/src/components/essentials/legend.ts:
```diff
@@ -80,6 +87,15 @@ export class Legend extends Component {
+					);
+				}
+			})
 			.style("fill", (d) => {
```
so we need both `class` & `fill`?

sophiiae replied:
we need `class` for color palette and `fill` for custom colors when user provides the color scale.

sophiiae commented on packages/core/src/model.ts:
```diff
@@ -584,47 +623,81 @@ export class ChartModel {
+		let pairingIndex = Tools.getProperty(colorPairingOptions, "index");
+		const availableColorPairings = {
+			"1-color": 4,
+			"2-color": 5,
+			"3-color": 5,
```
they have the same number of color pairs.

theiliad commented on packages/core/src/model.ts:
```diff
@@ -454,6 +465,46 @@ export class ChartModel {
+
+		if (
+			userProvidedScale == null ||
+			Object.keys(userProvidedScale).length < dataGroups.length
```
what if I'm using the color palette and I provide a color scale that changes just 1 color? Would that be disregarded here?

sophiiae replied:
Yes, if user only provides partial color scale, those colors will be assigned to the components repeatedly. Currently, we only apply colors either by class or color fill. I can make another PR later for this so user can partially override the color.

theiliad commented on packages/core/src/model.ts:
```diff
@@ -454,6 +465,46 @@ export class ChartModel {
 		return this.colorScale;
 	}
 
+	isCustomColorValid() {
```
```suggestion
	isUserProvidedColorScaleValid() {
```

theiliad commented on packages/core/src/model.ts:
```diff
@@ -1,8 +1,8 @@
 import * as Configuration from "./configuration";
 import { Tools } from "./tools";
-import * as colorPalettes from "./services/colorPalettes";
 import { Events, ScaleTypes } from "./interfaces";
+import { ColorClassNameTypes } from "./interfaces/enums";
```
```suggestion
import { Events, ScaleTypes, ColorClassNameTypes } from "./interfaces";
```
